File: bundle/edu.gemini.obslog/src/main/java/edu/gemini/obslog/instruments/GNIRSLogSegment.java
package edu.gemini.obslog.instruments;

import edu.gemini.obslog.config.model.OlLogItem;
import edu.gemini.obslog.core.OlSegmentType;
import edu.gemini.obslog.obslog.ConfigMap;
import edu.gemini.obslog.obslog.InstrumentLogSegment;
import edu.gemini.obslog.obslog.OlLogOptions;
import edu.gemini.shared.util.immutable.ImOption;
import edu.gemini.spModel.gemini.gnirs.GNIRSParams;

import java.util.List;
import java.util.logging.Logger;


public class GNIRSLogSegment extends InstrumentLogSegment {
    public static final Logger LOG = Logger.getLogger(GNIRSLogSegment.class.getName());

    private static final String NARROW_TYPE = "GNIRS";
    public static final OlSegmentType SEG_TYPE = new OlSegmentType(NARROW_TYPE);

    private static final String SEGMENT_CAPTION = "GNIRS Observing Log";

    private static final String WAVELENGTH_KEY = "wavelength";
    private static final String TYPE_KEY = "type";
    private static final String FILTER_KEY = "filter";
    private static final String CROSSDISPERSED_KEY = "crossDispersed";
    private static final String DISPERSER_KEY = "disperser";
    private static final String SLITWIDTH_KEY = "slitwidth";
    private static final String ACQMIRROR_KEY = "acqMirror";
    private static final String READMODE_KEY = "readMode";
    private static final String COADDS_KEY = "coadds";
    private static final String EXPOSURETIME_KEY = "exposureTime";
    private static final String PIXELSCALE_KEY = "pixelScale";
    private static final String CAMERA_KEY = "camera";

    public GNIRSLogSegment(final List<OlLogItem> logItems, final OlLogOptions obsLogOptions) {
        super(SEG_TYPE, logItems, obsLogOptions);
    }

    private void _decorateFilter(final ConfigMap map) {
        if (map == null) return;

        // Note that filterValue can return null here if there is no filter in the sequence.  Then this code needs
        // to divine it from other values based upon SeqExec details
        String filterValue = map.sget(FILTER_KEY);

        // This should work if filterValue == null - the check is added to keep
        // oldValueOf from complaining
        if (filterValue != null) {
            final GNIRSParams.Filter f = GNIRSParams.Filter.getFilter(filterValue, null);
            if (f != null) {
                final String typeValue = ImOption.apply(map.sget(TYPE_KEY)).getOrElse("");
                map.put(FILTER_KEY, !typeValue.equals("DARK") ? f.logValue() : "DARK");
                return;
            }
        }

        // If we make it here, we are divining the filter.
        final String crossValue = map.sget(CROSSDISPERSED_KEY);
        if (crossValue.equals("No")) {
            String lambdaValue = map.sget(WAVELENGTH_KEY);
            if (lambdaValue == null) return;
            try {
                double lambda = Double.parseDouble(lambdaValue);
                if (lambda < 1.17) {
                    filterValue = "X";
                } else if (lambda < 1.42) {
                    filterValue = "J";
                } else if (lambda < 1.86) {
                    filterValue = "H";
                } else if (lambda < 2.70) {
                    filterValue = "K";
                } else if (lambda < 4.30) {
                    filterValue = "L";
                } else {
                    filterValue = "M";
                }
            } catch (NumberFormatException ex) {
                filterValue = "?";
            }
        } else {
            filterValue = "XD";
        }

        // Finally, set the filter value for the not DARK and not in sequence section
        map.put(FILTER_KEY, filterValue);
    }

    // Note this replaces the "disperser" value with disperser/wavelength
    private void _decorateDisperser(final ConfigMap map) {
        if (map == null) return;

        final String value = map.sget(DISPERSER_KEY);
        if (value == null) return;
        final GNIRSParams.Disperser disperser = GNIRSParams.Disperser.getDisperser(value);

        final String wavelengthValue = map.sget(WAVELENGTH_KEY);
        if (disperser != null && wavelengthValue != null) {
            map.put(DISPERSER_KEY, disperser.logValue() + '/' + wavelengthValue);
        }
    }

    // Note this replaces the slitwidth value with the "log" value
    // Note at this time this doesn't bother with the decker the logs don't seem to either
    private void _decorateSlit(final ConfigMap map) {
        if (map == null) return;

        final String value = map.sget(SLITWIDTH_KEY);
        if (value == null) return;

        final GNIRSParams.SlitWidth s = GNIRSParams.SlitWidth.getSlitWidth(value);
        if (s != null) map.put(SLITWIDTH_KEY, s.logValue());
    }

    // Note this changes the acqmirror value from "in" to "Y"
    private void _decorateAcquisition(final ConfigMap map) {
        if (map == null) return;

        final String value = map.sget(ACQMIRROR_KEY);
        map.put(ACQMIRROR_KEY, (value != null && value.equals("in")) ? "Y" : "");
    }

    // This method looks at the coadds, readmode, and exposureTime to synthesize the new complicated exposure time
    private void _decorateExposureTime(final ConfigMap map) {
        if (map == null) return;

        final String readmodeValue = map.sget(READMODE_KEY);
        if (readmodeValue == null) {
            LOG.severe("No readMode found in GNIRS items");
        }

        final GNIRSParams.ReadMode rm = GNIRSParams.ReadMode.getReadMode(readmodeValue);
        final int lnrs = rm.getLowNoiseReads();

        final String coaddsValue = ImOption.apply(map.sget(COADDS_KEY)).getOrElse(() -> {
            LOG.severe("No coaddsValue found in GNIRS items");
            return "1";
        });

        // Note exposureTime is the entry that is reused for display
        final String exposureTimeValue = ImOption.apply(map.sget(EXPOSURETIME_KEY)).getOrElse(() -> {
            LOG.severe("No exposureTime property in GNIRS items");
            return "1";
        });

        map.put(EXPOSURETIME_KEY, exposureTimeValue + '/' + lnrs + '/' + coaddsValue);
    }

    // Note this replaces the camera item with the camera/prism value
    private void _decorateCamera(final ConfigMap map) {
        if (map == null) return;

        final String value = map.sget(WAVELENGTH_KEY);
        if (value == null) return;


        final String pscaleValue = map.sget(PIXELSCALE_KEY);
        final GNIRSParams.PixelScale ps = GNIRSParams.PixelScale.getPixelScale(pscaleValue);

        String camera;
        try {
            double lambda = Double.parseDouble(value);
            if (ps == GNIRSParams.PixelScale.PS_015) {
                    camera = (lambda > 2.5) ? "SR" : "SB";
            } else {
                    // pixel scale is PS_005
                    camera = (lambda > 2.5) ? "LR" : "LB";
            }
        } catch (NumberFormatException ex) {
            camera = "?";
        }

        // REL-1101: crossValue is either No, SXD, or LXD. Translate No into MIR.
        final String crossValue = map.sget(CROSSDISPERSED_KEY);
        final String prism = crossValue.equals("No") ? "MIR" : crossValue;
        map.put(CAMERA_KEY, camera + '/' + prism);
    }

    /**
     * Given an ObservationData object, create its possibly specialized bean data.  This method is a
     * factory for the particular segments type of bean data.
     *
     * @param map <code>UniqueConfigMap</code>
     */
    public void decorateObservationData(final ConfigMap map) {
        _decorateFilter(map);
        _decorateSlit(map);
        _decorateDisperser(map);
        _decorateExposureTime(map);
        _decorateAcquisition(map);
        _decorateCamera(map);
    }

    /**
     * Return the segment caption.
     *
     * @return The caption.
     */
    public String getSegmentCaption() {
        return SEGMENT_CAPTION;
    }
}

File: bundle/edu.gemini.obslog/src/main/scala/edu/gemini/obslog/instruments/GNIRSLogSegment.scala
package edu.gemini.obslog.instruments

import java.util.logging.Logger

import edu.gemini.obslog.config.model.OlLogItem
import edu.gemini.obslog.core.OlSegmentType
import edu.gemini.obslog.obslog.{ConfigMap, InstrumentLogSegment, OlLogOptions}
import edu.gemini.spModel.gemini.gnirs.{GNIRSParams, InstGNIRS}
import edu.gemini.spModel.obscomp.InstConstants

import GNIRSParams._
import GNIRSLogSegment._

import scalaz._
import Scalaz._

class GNIRSLogSegment(logItems: java.util.List[OlLogItem], obsLogOptions: OlLogOptions) extends InstrumentLogSegment(SegmentType, logItems, obsLogOptions) {

  override def decorateObservationData(map: ConfigMap): Unit = Option(map).foreach(m => {
    decorateFilter(m)
    decorateSlitWidth(m)
    decorateDisperser(m)
    decorateAcquisitionMirror(m)
    decorateCamera(m)
  })

  override def getSegmentCaption = SegmentCaption

  private def decorateFilter(map: ConfigMap): Unit = {
    // This is a hideous hack. I don't know how to improve it.
    def determineFilter: Option[String] = (for {
      v <- map.sgetOpt(FilterKey)
      f <- Option(Filter.getFilter(v, null))
    } yield map.sgetOpt(TypeKey).map {
      case InstConstants.DARK_OBSERVE_TYPE => InstConstants.DARK_OBSERVE_TYPE
      case _ => f.logValue()
    }).flatten

    def determineFilterFromWavelength: Option[String] = map.sgetOpt(CrossDispersedKey).flatMap { c =>
      if (yesOrNotNo(c)) "XD".some
      else map.sgetOpt(WavelengthKey).map { l => \/.fromTryCatchNonFatal {
          val lambda = l.toDouble
          FilterOrder.find(_.getMaxWavelength > lambda).getOrElse(GNIRSParams.Order.ONE).getBand
        }.getOrElse("?") }
      }

    // If we can determine the filter in one of the two ways, then adjust it.
    determineFilter.orElse(determineFilterFromWavelength).foreach(map.put(FilterKey, _))
  }

  // Replace the disperser value with disperser/wavelength.
  private def decorateDisperser(map: ConfigMap): Unit = for {
    v <- map.sgetOpt(DisperserKey)
    d <- Option(Disperser.getDisperser(v)).map(_.logValue())
    w <- map.sgetOpt(WavelengthKey)
  } map.put(DisperserKey, s"$d/$w")

  // This method looks at the coadds, readmode, and exposureTime to synthesize the new complicated exposure time
  private def decorateExposureTime(map: ConfigMap): Unit = {
    def lowNoiseReads: Option[Int] = {
      val lnrs = map.sgetOpt(ReadModeKey).map(ReadMode.getReadMode).map(_.getLowNoiseReads)
      if (lnrs.isEmpty) Log.severe("No readMode found in GNIRS items")
      lnrs
    }
    def extractWithLog(key: String, err: String, default: String): String = {
      val value = map.sgetOpt(key)
      if (value.isEmpty) Log.severe(err)
      value.getOrElse(default)
    }

    lowNoiseReads.foreach { lnrs =>
      val expTime = extractWithLog(ExposureTimeKey, "No exposureTime property in GNIRS items", "1")
      val coadds = extractWithLog(CoaddsKey, "No coaddsValue found in GNIRS items", "1")

      map.put(ExposureTimeKey, s"$expTime/$lnrs/$coadds")
    }
  }

  // Replace the slitwidth with the log value.
  private def decorateSlitWidth(map: ConfigMap): Unit = for {
    w <- map.sgetOpt(SlitWidthKey)
    s <- Option(SlitWidth.getSlitWidth(w))
  } map.put(SlitWidthKey, s.logValue())

  // This changes the acquisition mirror from "in" to "Y"; otherwise nothing.
  private def decorateAcquisitionMirror(map: ConfigMap): Unit = {
    val newValue = map.sgetOpt(AcqMirrorKey).collect {
      case "in" => "Y"
    }.getOrElse("")
    map.put(AcqMirrorKey, newValue)
  }

  private def decorateCamera(map: ConfigMap): Unit = {
    for {
      k <- map.sgetOpt(WavelengthKey)
      p <- map.sgetOpt(PixelScaleKey)
      c <- map.sgetOpt(CrossDispersedKey)
    } yield {
      val ps = PixelScale.getPixelScale(p)

      // Camera
      val cameraName = \/.fromTryCatchNonFatal {
        val lambda = k.toDouble
        ps match {
          case PixelScale.PS_015 if lambda > 2.5 => Camera.SHORT_RED
          case PixelScale.PS_015                 => Camera.SHORT_BLUE
          case _                 if lambda > 2.5 => Camera.LONG_RED
          case _                                 => Camera.LONG_BLUE
        }
      }.getOrElse("?")

      // Prism
      val prismName = yesOrNotNo(c) ? (ps match {
        case PixelScale.PS_015 => CrossDispersed.SXD
        case _                 => CrossDispersed.LXD
      }).displayValue() | "MIR"

      map.put(CameraKey, s"$cameraName/$prismName")
    }
  }
}

object GNIRSLogSegment {
  val Log = Logger.getLogger(classOf[GNIRSLogSegment].getName)
  val SegmentType = new OlSegmentType(InstGNIRS.SP_TYPE.narrowType)
  val SegmentCaption = "GNIRS Observing Log"

  val AcqMirrorKey      = "acqMirror"
  val CameraKey         = "camera"
  val CoaddsKey         = "coadds"
  val CrossDispersedKey = "crossDispersed"
  val DisperserKey      = "disperser"
  val ExposureTimeKey   = "exposureTime"
  val FilterKey         = "filter"
  val PixelScaleKey     = "pixelScale"
  val ReadModeKey       = "readMode"
  val SlitWidthKey      = "slitwidth"
  val TypeKey           = "type"
  val WavelengthKey     = "wavelength"

  // An order of filters to try.
  val FilterOrder = List(
    GNIRSParams.Order.SIX,
    GNIRSParams.Order.FIVE,
    GNIRSParams.Order.FOUR,
    GNIRSParams.Order.THREE,
    GNIRSParams.Order.TWO
  )

  // This is to handle the hideousness of cross dispersers.
  // Yes used to be a supported option, and now it is not: it is divided into SXD and LXD.
  // Thus, for backward compatibility, we determine if the cross disperser is yes (legacy) or not no.
  def yesOrNotNo(s: String): Boolean = s.equalsIgnoreCase("yes") || !s.equalsIgnoreCase("no")
}
